tests/e2e/iothub_e2e/aio/test_c2d.py

```diff
@@ -18,29 +19,65 @@
 class TestReceiveC2d(object):
     @pytest.mark.it("Can receive C2D")
     @pytest.mark.quicktest_suite
-    async def test_receive_c2d(self, client, service_helper, event_loop, leak_tracker):
+    async def test_receive_c2d(self, session_object, service_helper, event_loop, leak_tracker):
         leak_tracker.set_initial_object_list()
 
         message = json.dumps(get_random_dict())
 
-        received_message = None
-        received = asyncio.Event()
+        queue = asyncio.Queue()
 
-        async def handle_on_message_received(message):
-            nonlocal received_message, received
-            logger.info("received {}".format(message))
-            received_message = message
-            event_loop.call_soon_threadsafe(received.set)
+        async def listener(sess):
+            try:
+                async with sess.messages() as messages:
+                    async for message in messages:
+                        await queue.put(message)
+            except asyncio.CancelledError:
+                # this happens during shutdown. no need to log this.
+                raise
+            except BaseException:
+                # Without this line, exceptions get silently ignored until
+                # we await the listener task.
+                logger.error("Exception", exc_info=True)
+                raise
 
-        client.on_message_received = handle_on_message_received
-        await client.start_message_receive()
+        async with session_object:
+            listener_task = asyncio.create_task(listener(session_object))
 
-        await service_helper.send_c2d(message, {})
+            await service_helper.send_c2d(message, {})
 
-        await asyncio.wait_for(received.wait(), 60)
-        assert received.is_set()
+            received_message = await queue.get()
 
-        assert received_message.data.decode("utf-8") == message
+        assert session_object.connected is False
+        with pytest.raises(asyncio.CancelledError):
+            await listener_task
+        listener_task = None
 
-        received_message = None  # so this isn't tagged as a leak
+        assert received_message.payload == message
+
+        del received_message
+        leak_tracker.check_for_leaks()
+
+    @pytest.mark.it("Can receive C2D using anext")
```

Comment:
There must be a difference since one leaks and the other doesn't.  Honestly, that surprises me.

I can't find the thread right now, but there is a difference between `__anext__` and `anext`. I'm thinking maybe we should test both.

Reply:
That is a deeply unpleasant revelation.